src/webgpu/api/validation/encoding/cmds/render/dynamic_state.spec.ts

```diff
@@ -130,71 +130,97 @@ class F extends ValidationTest {
-        { attachmentWidth: 1024, attachmentHeight: 1 },
-        { attachmentWidth: 1, attachmentHeight: 1024 },
-      ])
-      .combineWithParams([
-        // Control case: a full viewport is valid.
-        { dx: 0, dy: 0, dw: 0, dh: 0 },
-
-        // Other valid cases with a partial viewport.
-        { dx: 1, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: -1 },
-        { dx: 0, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: -1 },
-
-        // Test with a small value that causes the viewport to go outside the attachment.
-        { dx: 1, dy: 0, dw: 0, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: 0 },
-        { dx: 0, dy: 0, dw: 1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: 1 },
-      ])
+    u.combineWithParams([
+      // Control case: max viewport is valid.
+      { mx: 0, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+
+      // Other valid cases
+      { mx: -1, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: -2, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -1, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -2, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 0, dy: 1, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 0, dy: 0, dw: -1, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 0, dw: 0, dh: -1 },
+
+      // Cases that go outside the allowed bounds
+      { mx: -2, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -2, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 1, dy: 0, dw: -1, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 1, dw: 0, dh: -1 },
+
+      // Cases that exceed the max viewport size
+      { mx: 0, my: 0, dx: 0, dy: 0, dw: 1, dh: 0 },
+      { mx: 0, my: 0, dx: 0, dy: 0, dw: 0, dh: 1 },
+    ])
   )
   .fn(t => {
-    const { attachmentWidth, attachmentHeight, dx, dy, dw, dh } = t.params;
-    const x = dx;
-    const y = dy;
-    const w = attachmentWidth + dw;
-    const h = attachmentWidth + dh;
-
-    const success = x + w <= attachmentWidth && y + h <= attachmentHeight;
-    t.testViewportCall(
-      success,
-      { x, y, w, h, minDepth: 0, maxDepth: 1 },
-      { width: attachmentWidth, height: attachmentHeight, depthOrArrayLayers: 1 }
-    );
+    const { mx, my, dx, dy } = t.params;
+    const maxViewportSize = t.device.limits.maxTextureDimension2D;
+    const maxViewportBounds = maxViewportSize * 2;
+
+    const x = maxViewportSize * mx + dx;
+    const y = maxViewportSize * my + dy;
+    const w = maxViewportSize;
+    const h = maxViewportSize;
+
+    const inBounds =
+      x >= -maxViewportBounds &&
+      y >= -maxViewportBounds &&
+      x + w <= maxViewportBounds - 1 &&
```

Comment:
We should have some fractional test cases (x/y/w/h are floats). In particular it would be easy for an implementation to incorrectly check `x + w < maxViewportBounds` instead of `x + w <= maxViewportBound - 1`.

Reply:
Done